### ISSUE TITLE
Network disconnects when switching between wifi and cellular

### ISSUE THREAD

**Issue by jmencak** [t895/DNSNet#8]
### Issue description
When switching between wifi and celular, I get a complete disconnect from the network.

### DNSNet version
F-Droid, 1.0.14

### Device
Google Pixel 8

### OS
GrapheneOS, Android 15

### Additional information
1) I love this app.  The UI is clean and it does exactly what I want this app to do and nothing more.  You nailed it, thank you.
2) Rethink (DNS + Firewall + VPN) and personalDNSfilter DNS does not have this issue on my setup.
3) I see the following in the logs, but it might not be relevant:

```
12-19 08:23:03.390 15670 18793 I k       : java.lang.ClassCastException: W3.y0 cannot be cast to W3.z2
12-19 08:23:03.390 15670 18793 I k       :      at V2.k.a(SourceFile:28)
12-19 08:23:03.390 15670 18793 I k       :      at V2.f.i(SourceFile:28)
12-19 08:23:03.390 15670 18793 I k       :      at V2.f.e(SourceFile:282)
12-19 08:23:03.390 15670 18793 I k       :      at V2.f.j(SourceFile:50)
12-19 08:23:03.390 15670 18793 I k       :      at V2.f.run(SourceFile:42)
12-19 08:23:03.390 15670 18793 I k       :      at java.lang.Thread.run(Thread.java:1117)
```

**Comment by t895** [comment]
Hmn, I'll try to replicate this issue in my end
I can't seem to do it on my first try though

I'm also working on another issue where queries can be slow to resolve. Could you try just waiting a longer period of time for the site to load during a network switch?
If it's just slow, I should have a fix out relatively soon.

**Comment by jmencak** [comment]
> I'm also working on another issue where queries can be slow to resolve. Could you try just waiting a longer period of time for the site to load during a network switch?

Thank you for looking into this!  I tried waiting for 20 minutes and the network switch never happened.  I had to go to the DNSNet application, stop it using the stop button and start it again using the play button.

**Comment by lukay101** [comment]
I also have a problem that dnsnet doesn't seem to work on some apps like Grindr when changing networks. I have to disable dns net and enable again for the app to be able to connect again.

**Comment by t895** [comment]
Sorry for the awkward wait on this. I spend a bit of last night working on a fix for this. Turns out Android's `NetworkCallback` has some odd behavior that I wasn't accounting for that would prevent the VPN from restarting during a network change.

Assuming Android doesn't throw me any curveballs (it will), I'll have a fix out by the end of today.

**Comment by MrQubo** [comment]
I also have the same problem on GrapheneOS, Android 15. Maybe it's GrapheneOS-specific issue?

**Comment by t895** [comment]
@MrQubo I use GrapheneOS on my primary device and while it happens there, I verified that it also happens on stock emulator system images from Google as well

I just pushed a fix to the repo that fixes the issue in both instances

Edit - Just to be clear, I did not find a GrapheneOS specific problem

**Comment by lukay101** [comment]
I'm on oneui by Samsung OS

**Comment by t895** [comment]
I just pushed an update to the GitHub releases page. Please let me know if it was fixed on your end.

**Comment by lukay101** [comment]
> I just pushed an update to the GitHub releases page. Please let me know if it was fixed on your end.

Thank you, seems like my problem with the specific app after switching to mobile data is solved🎉 will be testing further

**Comment by jmencak** [comment]
> I just pushed an update to the GitHub releases page. Please let me know if it was fixed on your end.

I can confirm that switching between wifi/mobile works fine on fully up-to-date GrapheneOS 15, Google Pixel 8. Thank you!

**Comment by MrQubo** [comment]
Works for me as well. Thanks for your work!

**Comment by lukay101** [comment]
> I just pushed an update to the GitHub releases page. Please let me know if it was fixed on your end.

Will you push this update to fdroid too?

**Comment by michaelblyons** [comment]
> Will you push this update to fdroid too?

It's built on F-Droid now. Will probably get distributed in a few days.
